### PR TITLE
[web] Fix disposal of browser history on hot restart

### PR DESCRIPTION
**Pull request by mdebbar**
If we don't dispose of `BrowserHistory` instances, they keep listening to `popstate` events even after a hot restart. So we end up with duplicate listeners.

In this PR, we make sure to always dispose of `BrowserHistory` instance but without resetting the browser history entries (to keep the effect of [this PR](https://github.com/flutter/engine/pull/27872)).

Fixes https://github.com/flutter/flutter/issues/90031
Fixes https://github.com/flutter/flutter/issues/83864

### CODE REVIEW

File: lib/web_ui/test/engine/history_test.dart [hunk 1]
```diff
@@ -326,6 +381,54 @@ void testMain() {
         // TODO(mdebbar): https://github.com/flutter/flutter/issues/50836
         skip: browserEngine == BrowserEngine.edge);
 
+    test('disposes of its listener without touching history', () async {
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/test/engine/history_test.dart:384]: They are slightly different. One for single entry and one for multi entry.

File: lib/web_ui/lib/src/engine/navigation/history.dart [hunk 2]
```diff
@@ -199,8 +214,7 @@ class MultiEntriesBrowserHistory extends BrowserHistory {
     if (_isDisposed || urlStrategy == null) {
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/navigation/history.dart:214]: Nice catch! Let me use a separate boolean `_isTornDown`.

File: lib/web_ui/test/engine/history_test.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (chunhtai) [reply lib/web_ui/test/engine/history_test.dart:392]: ah gotcha

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by cdprete** [comment]
Hi.
I'm actually experiencing the issue fixed here with Flutter 2.5.3.
Flutter doctor output is at [#94675](https://github.com/flutter/flutter/issues/94675)

**Pull request merged**